File: eyan_helper/src/main/java/eu/eyan/util/io/CachedFileLineReader.java
package eu.eyan.util.io;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.charset.Charset;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.google.common.collect.Lists;

import eu.eyan.log.Log;
import eu.eyan.util.collection.MapsPlus;

/**
 * Only the start offsets going to be stored in an arraylist. 200 lines are
 * stored in a map-cache.
 */
public class CachedFileLineReader implements Iterable<String> {

	private static final int LINE_COUNT_EARLY_READ = 100;

	private FileChannel fileChannel;

	private FileInputStream fileInputStream;

	protected List<long[]> lineOffsets = Lists.newArrayList();

	protected Map<Integer, String> lineCache = MapsPlus.newMaxSizeHashMap(Runtime.getRuntime().availableProcessors() * LINE_COUNT_EARLY_READ);

	private String longestLine = "";

	private long fileLength;

	public synchronized void close() {
		try {
			if (fileInputStream != null) {
				fileInputStream.close();
			}
			if (fileChannel != null) {
				fileChannel.close();
			}
			lineOffsets.clear();
			lineCache.clear();
			longestLine = "";
			fileLength = 0;
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	public void load(File file, Consumer<Integer> loadFileProgressChangedEvent) {
		synchronized (lineOffsets) {
			try {
				fileLength = file.length();
				System.out.println("Loading " + file + " " + fileLength + " bytes");
				FileLineStartOffsetReader lnr = new FileLineStartOffsetReader(file);
				long endIndex = 0;
				long startOffset = 0;
				long progressPercent = 0;
				int longestLineIndex = 0;
				long longestLineLength = 0;
				while (startOffset != -1) {
					startOffset = lnr.readLine();
					if (startOffset != -1) {
						endIndex = lnr.getOffset();
						lineOffsets.add(new long[] { startOffset, endIndex });
						if (endIndex - startOffset > longestLineLength) {
							longestLineIndex = lineOffsets.size() - 1;
							longestLineLength = endIndex - startOffset;
						}
					}

					final long newProgressPercent = fileLength == 0 ? 100 : endIndex * 100 / fileLength;
					if (progressPercent != newProgressPercent && loadFileProgressChangedEvent != null) {
						loadFileProgressChangedEvent.accept(Math.toIntExact(newProgressPercent));
						progressPercent = newProgressPercent;
					}
				}
				lnr.close();

				fileInputStream = new FileInputStream(file);
				fileChannel = fileInputStream.getChannel();
				longestLine = get(longestLineIndex);
				// if (fileLength != endIndex) { //special characters brake the
				// offsets
				// System.out.println("Length: " + fileLength + " endOffset:" +
				// endIndex + " ");
				// System.err.println("Error at loading file. There are newline problems! ");
				// }
				// System.gc();// wow... Test for big files again if it is
				// really necessary
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}

	public String getLongestLine() {
		return longestLine;
	}

	public int size() {
		synchronized (lineOffsets) {
			return lineOffsets.size();
		}
	}

	public String get(int index) {
		String line = lineCache.get(index);
		if (line == null) {
			try {
				synchronized (lineOffsets) {
					for (int i = index; i < lineOffsets.size() && i < index + LINE_COUNT_EARLY_READ; i++) {
						String nextLine = readFromFile(i);
						lineCache.put(i, nextLine);
						if (i == index) {
							line = nextLine;
						}
					}
				}
			} catch (IOException e) {
				Log.error(e);
			}
		}
		return line;
	}

	private String readFromFile(int index) throws IOException {
		long start = lineOffsets.get(index)[0];
		long end = lineOffsets.get(index)[1];
		ByteBuffer dst = ByteBuffer.allocate(Math.toIntExact(end - start));
		fileChannel.read(dst, start);
		String line = new String(dst.array(), Charset.forName("UTF-8"));
		return line;
	}

	@Override
	public Iterator<String> iterator() {
		return new Iterator<String>() {
			private int index = 0;

			@Override
			public boolean hasNext() {
				synchronized (lineOffsets) {
					return index < lineOffsets.size();
				}
			}

			@Override
			public String next() {
				if(size() != 0) Log.debug("Line "+(index+1)+" "+(100*(index+1)/size())+"%");
				String line = get(index);
				index++;
				return line;
			}
		};
	}

	public Matcher findFirst(String pattern) {
		Matcher matcher = Pattern.compile(pattern).matcher("");
		for (String line : this) {
			line = line.replaceAll("\r\n", "");
			matcher.reset(line);
			if (matcher.matches()) {
				return matcher;
			}
		}
		return null;
	}
}

File: eyan_helper/src/main/scala/eu/eyan/util/io/CachedFileLineReader.scala
package eu.eyan.util.io

import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.channels.FileChannel
import java.nio.charset.Charset
import java.util.regex.Pattern

import eu.eyan.log.Log
import eu.eyan.util.collection.MapsPlus._
import java.io.Closeable
import eu.eyan.util.string.StringPlus.StringPlusImplicit
import scala.collection.mutable.ListBuffer
import eu.eyan.util.scala.Try
import eu.eyan.util.scala.TryFinallyClose
import eu.eyan.util.java.lang.RuntimePlus._

object CachedFileLineReader {
  def apply(file: String) = new CachedFileLineReader().load(file.asFile)
  def apply(file: File) = new CachedFileLineReader().load(file)
  def apply() = new CachedFileLineReader()
}

class CachedFileLineReader extends Iterable[String] with Closeable {

  private val LINE_COUNT_EARLY_READ = 100
  protected var lineOffsets = Vector[Array[Long]]()
  var longestLineIndex = 0

  def getLineOffsets = lineOffsets // FIMXE getter for java? for sortable
  def getLineOffsets(idx: Int) = lineOffsets(idx) // FIMXE getter for java? for sortable

  // FIXME must be thread safe!
//  private val lineCache = maxSizeMutableMap[Int, String](availableProcessors * LINE_COUNT_EARLY_READ) 
  private var lineCache = maxSizeImmutableMap[Int, String](availableProcessors * LINE_COUNT_EARLY_READ) 
  val NL = "\r\n"

  private var fileChannel: FileChannel = null
  private var fileInputStream: FileInputStream = null

  private def readFromFile(index: Int) = {
    val startEnd = getLineStartEnd(index)
    val start = startEnd(0)
    val end = startEnd(1)
    val byteBuffer = allocateBuffer(start, end)
    fileChannelRead(byteBuffer, start)
    val line = createLine(byteBuffer)
    line
  }

  private def readFromFile(startLineIndex: Int, endLineIndex: Int) = {
    val lastIndex = Math.min(endLineIndex, lineOffsets.size - 1)
    val start = getLineStartEnd(startLineIndex)(0)
    val end = getLineStartEnd(lastIndex)(1)

    val byteBuffer = allocateBuffer(start, end)
    fileChannelRead(byteBuffer, start)
    val linesArray = byteBuffer.array
    val lines = for (index <- startLineIndex to lastIndex) yield {
      val lineArray = linesArray.slice((lineOffsets(index)(0) - start).toInt, (lineOffsets(index)(1) - start).toInt)
      val line = createLine(lineArray)
      (index, line)
    }
    lines
  }

  def getLineStartEnd(index: Int) = lineOffsets(index) // FIXME slow ?
  def allocateBuffer(start: Long, end: Long) = ByteBuffer.allocate(Math.toIntExact(end - start))
  def fileChannelRead(byteBuffer: ByteBuffer, start: Long) = fileChannel.read(byteBuffer, start)
  def createLine(byteBuffer: ByteBuffer): String = createLine(byteBuffer.array)
  def createLine(array: Array[Byte]) = new String(array, Charset.forName("UTF-8"))

  def get(index: Int) = {
    val line = lineCache.get(index)
    if (line.isEmpty) Try {
      lineOffsets.synchronized {
        lineCache ++= readFromFile(index, index + 1 + LINE_COUNT_EARLY_READ)
      }
    }

    lineCache.get(index)
  }

  def load(file: File, loadFileProgressChangedEvent: Int => Unit = i => {}) = {
    close
    val fileLength = file.length

    Log.info("Loading " + file + " " + fileLength + " bytes")

    var progressPercent = 0L
    var longestLineLength = 0L
    var endIndex = 0L

    //TODO exchange to observable
    def readerCallback(index: Int, startOffset: Long, endOffset: Long) = {
      if (endOffset - startOffset > longestLineLength) {
        longestLineIndex = index
        longestLineLength = endOffset - startOffset
      }
      val newProgressPercent = if (fileLength == 0) 100 else endOffset * 100 / fileLength
      if (progressPercent != newProgressPercent && loadFileProgressChangedEvent != null) {
        loadFileProgressChangedEvent(Math.toIntExact(newProgressPercent))
        progressPercent = newProgressPercent
      }
      endIndex = endOffset
    }

    TryFinallyClose(FileLineStartOffsetReader(file), { lnr: FileLineStartOffsetReader => { lineOffsets = lnr.iterator(readerCallback).toVector } })

    fileInputStream = new FileInputStream(file)
    fileChannel = fileInputStream.getChannel()
    if (fileLength != endIndex) {
      // special characters brake the offsets
      Log.error("special characters brake the offsets: " + file.getAbsolutePath)
      Log.error("Length: " + fileLength + " endOffset:" + endIndex + NL + "Error at loading file. There are newline problems! ")
    }

    this
  }

  def close = Try {
    CloseablePlus.closeQuietly(fileInputStream, fileChannel)
    lineOffsets = Vector[Array[Long]]()
    lineCache = Map()
  }

  def getLongestLine = get(longestLineIndex)

  override def size = lineOffsets.synchronized { lineOffsets.size }

  def lines = iterator
  def iterator = new Iterator[String] {
    private var index = 0
    def hasNext = lineOffsets.synchronized { index < lineOffsets.size }
    def next = {
      if (lineOffsets.size != 0) Log.debug("Line " + (index + 1) + " " + (100 * (index + 1) / lineOffsets.size) + "%")
      val line = get(index)
      index += 1
      line.get
    }
  }

  def findFirst(pattern: String) = {
    val matcher = Pattern.compile(pattern).matcher("")

    val first = this.iterator.find(s => { matcher.reset(s.replaceAll(NL, "")); matcher.matches })

    if (first.nonEmpty) {
      val m = Pattern.compile(pattern).matcher(first.get.replaceAll(NL, ""))
      m.matches
      m
    } else null
  }
}